Managers/MapManager.gd

```diff
@@ -51,27 +51,162 @@ func create_map(map_floors_width: Array[int] = map_width_array) -> MapBase:
+	## since they could be the only destinations of a room from the previous floor (happens when current floor's size [br]
+	## is smaller or equal than previous)
+	var _neighbors_events: Array[String] = []
+	var delta_xs: Array[int] = []
+
+	# If rightmost room of floor
+	if current_room.room_position.x == map.rooms[current_room.room_position.y].size() - 1 or map.rooms[current_room.room_position.y][current_room.room_position.x+1] == null:
+		delta_xs = [-1, 0]
+	else:
+		delta_xs = [-2, -1, 0]
+		
+	# Get the neighbor and current rooms types
+	for delta_x: int in delta_xs:
+		var _neighbor_x: int = current_room.room_position.x + delta_x
+		var _neighbor_y: int = current_room.room_position.y
+		if _neighbor_x >= 0 && _neighbor_x < map.rooms[_neighbor_y].size() && map.rooms[_neighbor_y][_neighbor_x] != null:
+			_neighbors_events.append(map.rooms[_neighbor_y][_neighbor_x].room_event.get_event_name())
+
+	# If our current room has no neighbors on the left, ignore this rule
+	if _neighbors_events.size() >= 2:
+		var _unique: Array[String] = []
+
+		# Remove duplicates
+		for event in _neighbors_events:
+			if not _unique.has(event):
+				_unique.append(event)
+		
+		# If there is only one event type among all rooms, redraw event
+		if _unique.size() == 1:
+			return false
+
+
+	## Rule 4: No Heal rooms two floors before Boss
+	if (current_room_event_name == "heal" && current_room.room_position.y == map.rooms.size() - 3):
+		return false
+
+	return true
+
+
+## Draws room type randomly, with each type having a set probability [br]
+## Algorithm taken from this: https://stackoverflow.com/a/1761646
+func draw_room_type() -> EventBase:
+	var sum: int = 0
+	for proba in GlobalVar.EVENTS_PROBABILITIES:
+		sum += proba
+	var _rand_type_number: int = randi_range(0, sum-1)
+
+	for _rand_type_index: int in range(GlobalVar.EVENTS_CLASSIFICATION.size()):
+		if _rand_type_number < GlobalVar.EVENTS_PROBABILITIES[_rand_type_index]:
+			return GlobalVar.EVENTS_CLASSIFICATION[_rand_type_index].new()
+
+		_rand_type_number -= GlobalVar.EVENTS_PROBABILITIES[_rand_type_index]
+	
+	# This shouldn't be reached
+	return null
+
+## Assign events to existing rooms with set probabilities.
+func assign_events(map: MapBase = current_map) -> void:
+	# Scan the whole map, assign events to valid rooms
+	for index_height: int in range(map.rooms.size()):
+		for index_width: int in range(map.rooms[index_height].size()):
+			# If not a room, ignore
+			if map.rooms[index_height][index_width] == null:
+				continue
+			
+			var _current_room: RoomBase = map.rooms[index_height][index_width]
+
+			# Set fixed rooms here
+			if index_height == map.rooms.size() - 2:
+				# Rooms before boss are Heal rooms
+				_current_room.room_event = EventHeal.new()
+
+			elif index_height == map.rooms.size() - 1:
+				# Boss room, TO BE ASSIGNED
+				_current_room.room_event = EventMob.new()
+
+			else:
+				while not is_room_event_correct(_current_room, map):
```

Comment:
are we sure this will end ? i think testing how long this takes could be good
would it be possible to end with a map distribution such as for a given room, there will always be one of the rule that can't be upheld ?

Reply:
From my quick testing (just printing when a rule is not respected and the event is picked again), it usually picks a room after 0 to 2 attemps, so it does end pretty quickly, because except Rule 3, the rest of the rules apply mostly to Heal rooms. 
So technically, a whole map with only Mobs, Random and Dialogue upholds all rules (e.g. M,R,D,M,R,D,... repeat for each floor)
It's obviously not what we want, but is ensures that the generation does end.

Reply:
does the % of each event stay stable / close to target after changing due to rules ?

Reply:
![image](https://github.com/user-attachments/assets/2631a65e-ca28-4626-a53d-8cc45685bd25)
![image](https://github.com/user-attachments/assets/8a51ec0f-d450-4f5a-8dea-06502d1a62b6)

We usually end up with less Shop and Heals rooms since those are the most restricted, resulting in more of the rest

Reply:
Actually 2 shops out of 61 rooms is not a lot yeah, we could increase probabilities to adjust or something else

Reply:
A few things:
- printing the target % next to the actual % could be nice to quickly know how far away we are
- we can adjust % ofc
- maybe you should add some boundaries on the actual % for each event (lower and upper bound), and generate the map again if the actual % are too far off from the target

Reply:
With boundaries, won't we get a case where the map generation never ends from being very unlucky? It'll depend on how strict we are with boundaries ofc, I'll write the map generation time test.

Also i'm stupid, for shops 5% out of 61 is still 3 rooms, so it is close to what we want

Reply:
![image](https://github.com/user-attachments/assets/43bede4e-f5d1-4f2a-b2ee-592889fa00a4)
![image](https://github.com/user-attachments/assets/71d96195-6b6e-4101-be01-0b0bcae1f4bc)
![image](https://github.com/user-attachments/assets/0a32e230-0059-43b6-b4c6-bdd6d1bc807b)

Just some more generations, but yeah it does fluctuate quite a bit
I'll implement the changes you've pointed out and we can see again

Reply:
it indeed seems to fluctuate a lot
To change that, maybe we can adjust the probabilities of a given type being chosen given the amount of each room type already select ? The actual math to do that might be a bit complex, but we might be able to eyeball it

Reply:
We can try that, I get the idea but don't have the implementation in mind at all, it's a bit late rn 

variable probabilities based on what has been picked vs how much we expect for each event, no need of the exact number everytime we generate the map, while respecting the rules

Reply:
I think you can do this without needing too much code changes (there will be code added though)
I'm thinking of making a copy of the original % array / dictionary, and modify the values inside it as we go
For how much should be removed / added i'm still trying of a good way to do it (just the difference between target and actual seems a bit too harsh, especially when there are few events selected at the start)

Reply:
Could be a sort of priority system, a "what we need most".

Also small note: in the actual implementation, the probabilities don't need to be normalized or add to 100 for it to work. Just to say that if we increase only one value, it won't break. Normalizing is still better for understanding ofc

Reply:
yes i saw that it doesn't need to sum up to 100, we can normalize, or we don't, it doesn't really matter since it's internal